src/Connection.php

```diff
@@ -810,10 +810,8 @@ public function executeCacheQuery(string $sql, array $params, array $types, Quer
                 $value = [];
             }
 
-            if (isset($value[$realKey])) {
-                [$columnNames, $rows] = $value[$realKey];
-
-                return new Result(new ArrayResult($columnNames, $rows), $this);
+            if (isset($value[$realKey]) && $value[$realKey] instanceof ArrayResult) {
+                return new Result($value[$realKey], $this);
```

Comment:
The changes made here in `Connection` introduced a https://github.com/doctrine/dbal/labels/BC%20Break

Before this PR, the values in the cache were uncoupled from the `ArrayResult`.
Since this PR, the cache and the `Result`/`ArrayResult` are now coupled: hitting the `\Doctrine\DBAL\Result::free` method erases all the cache content as well :scream:

Reply:
Please file a bug report with clear steps to reproduce.